dags/property_definitions.py

```diff
@@ -1,273 +1,299 @@
+        ).result()
+        return TimeRange(start_time, end_time)
+
+
+@dagster.op
+def setup_job(
+    cluster: dagster.ResourceParam[ClickhouseCluster],
+    config: PropertyDefinitionsConfig,
+) -> TimeRange:
+    """Validates the job configuration to be provided to other ops."""
+    return config.validate(cluster)
+
+
+@dataclass(frozen=True)
+class DetectPropertyTypeExpression:
+    source_column: str
+
+    def __str__(self) -> str:
+        # largely derived from https://github.com/PostHog/posthog/blob/052f4ea40c5043909115f835f09445e18dd9727c/rust/property-defs-rs/src/types.rs#L314-L373
+        return f"""
+            arrayMap(
+                (name, value) -> (name, multiIf(
+                    -- special cases: key patterns
+                    name ilike 'utm_%', 'String',
+                    name ilike '$feature/%', 'String',
+                    name ilike '$feature_flag_response', 'String',
+                    name ilike '$survey_response%', 'String',
+                    -- special cases: timestamp detection
+                    (
+                        multiSearchAnyCaseInsensitive(name, ['time', 'timestamp', 'date', '_at', '-at', 'createdat', 'updatedat'])
+                        AND JSONType(value) IN ('Int64', 'UInt64', 'Double')
+                        AND JSONExtract(value, 'Nullable(Float)') >= toUnixTimestamp(now() - interval '6 months')
+                    ), 'DateTime',
+                    -- special cases: string value patterns
+                    (
+                        JSONType(value) = 'String'
+                        AND trimBoth(JSONExtractString(value)) IN ('true', 'TRUE', 'false', 'FALSE')
+                    ), 'Boolean',
+                    (
+                        JSONType(value) = 'String'
+                        AND length(trimBoth(JSONExtractString(value)) as trimmed_value) >= 10  -- require at least a date part
+                        AND parseDateTime64BestEffortOrNull(trimmed_value) IS NOT NULL  -- can be parsed as a date
+                        AND JSONExtract(trimmed_value, 'Nullable(Float)') IS NULL  -- but not as a timestamp
+                    ), 'DateTime',
+                    -- primitive types
+                    JSONType(value) = 'Bool', 'Boolean',
+                    JSONType(value) = 'String', 'String',
+                    JSONType(value) IN ('Int64', 'UInt64', 'Double'), 'Numeric',
+                    NULL
+                )),
+                JSONExtractKeysAndValuesRaw({self.source_column})
+            )
+        """
+
+
+COMMON_PROPERTY_FILTERS = """
+-- https://github.com/PostHog/posthog/blob/052f4ea40c5043909115f835f09445e18dd9727c/rust/property-defs-rs/src/types.rs#L17-L28
+name NOT IN ('$set', '$set_once', '$unset', '$group_0', '$group_1', '$group_2', '$group_3', '$group_4', '$groups')
+-- https://github.com/PostHog/posthog/blob/052f4ea40c5043909115f835f09445e18dd9727c/rust/property-defs-rs/src/types.rs#L279-L286
+AND length(name) <= 200
+"""
```

Comment:
This could probably be included in the type detection expression above to make it a bit more robust